### ISSUE TITLE
Cannot use Surge due to Fatal Surge Startup Error

### ISSUE THREAD

**Issue by NuloSois** [surge-synthesizer/surge#7544]
**Bug Description:**
Fatal Surge Startup Error, failed to retrieve known folder path. Folder attempt is Documents. Error code is -2147024894

**Surge XT Version**
This information is found on the About screen, which you get to from the bottom right menu.

* Version: 1.3.1
* Plugin Type: [VST2, VST3, AU]
* Bitness: [32-bit or 64-bit]

**Reproduction Steps:**
1. Open FL Studio
2. Go to Add Plugins -> More Plugins
3. Select and Add Surge XT
4. Get Screenshot

**Expected Behavior:**
I expected Surge XT to either work normally or have the same errors I was having with 1.2.3 (being unable to edit Surge projects after loading).

**Screenshots:**
![image](https://github.com/surge-synthesizer/surge/assets/104397117/7e547552-03cc-407f-98c9-2599f76114ee)

**Computer Information (please complete the following!):**
 - OS:  Windows 11
 - Host: FL Studio
 - Version: 12

**Additional Information:**

**Comment by mkruselj** [comment]
We've already had a couple of such reports, and in both cases OneDrive was at fault, somehow Documents folder ended up being inside OneDrive folder and then there were some permissions issues IIRC.

**Comment by NuloSois** [comment]
Yeah, I noticed OneDrive is being a huge proble to me, do you have any expectation to solve this bug on the next release or something alike?

**Comment by baconpaul** [comment]
this isn't something we can solve its something you have to solve on your system. Basically we are asking the OS for a 'Documents" folder and there isn't one.

https://superuser.com/questions/1725456/wha-cant-i-move-my-documents-folder I think this is the document you need.

but basically you need to make your window system answer the windows api for the documents folder location again. (I use mac but if you are on our discord we've been through this a few times, and we should write a FAQ too!)

**Comment by parduz** [comment]
1) couldn't be possible to allow the users, as an option, to manually set the data folders so they could be anywhere?
2) is it unavoidable to be a fatal error? could'nt Surge just open using the factory data?

**Comment by mkruselj** [comment]
That option to set the user data paths already exists.

User data contains the patch database and the prefs file which both need read and write access, so if there's no access to Documents folder, it cannot NOT be a fatal error.

**Comment by parduz** [comment]
> That option to set the user data paths already exists, but can only be set when Surge is running properly.
> User data contains the patch database and the prefs file which both need read and write access, so if there's no access to Documents folder, it cannot NOT be a fatal error.

So, if default user data path is not accessible it is impossible to set it somewhere else 'cause Surge only runs if it is.
Seems a dead loop, to me.

**Comment by mkruselj** [comment]
Oh actually, on Windows one can use a symbolic link (there is a checkbox to create a junction to user folder next to the plugin file) to point Surge to where the user folder is.

**Comment by parduz** [comment]
> Oh actually, on Windows one can use a symbolic link (there is a checkbox to create a junction to user folder next to the plugin file) to point Surge to where the user folder is.

But the patches then will be writeable?

I've symlinked all the factory and 3rdparty patches on my user folder: the patches sounds well but i can't overwrite them if i make changes

**Comment by mkruselj** [comment]
I guess it depends on the write permissions of the folder that is the target of the symbolic link...

**Comment by parduz** [comment]
> I guess it depends on the write permissions of the folder that is the target of the symbolic link

Exactly. Here problem is not being able to open SurgeXT 'cause user folder is not accessible.

- problem is the user folder is not accessible
- you said that on Windows one can use a symbolic link to point Surge to where the user folder is.
- i said that perhaps the patch then would have been read but not writable, so the problem still exists.

Unless you wanted to say that in order to setup a custom user data folder, when Surge is not opening 'cause the default one in not accessible, one have to:

- create a symlink so that Surge at least opens
- open Surge
- setup a custom path.

**Comment by mkruselj** [comment]
Bottom line is, the Documents folder should not be inside OneDrive folder, else there is trouble. Nothing we can do about this.

**Comment by parduz** [comment]
While this all is true, i think Surge should be less harsh in his reaction. Closing itself may results in DAW crashing to desktop.
At least popping out a textbox to write an accessible path for the user data, but why couldn't it live with just the installed folders?

**Comment by mkruselj** [comment]
I explained why it cannot live with just the installed folders: the patch database requires read/write access and can as such only be in the user data folder, which is in Documents by default. If OS doesn't give us a valid path to Documents, we cannot do anything about it.

**Comment by baconpaul** [comment]
> While this all is true, i think Surge should be less harsh in his reaction. Closing itself may results in DAW crashing to desktop. At least popping out a textbox to write an accessible path for the user data, but why couldn't it live with just the installed folders?

I would happily merge a change to improve this. Our documentation explains how to build and develop the synth!

this broken windows condition crashed the entire synth until 1.2 and we placed the message on screen. But the code is not defensive against a missing documents folder so it is quite a bit of work to fix for MIs configured windows environments

**Comment by baconpaul** [comment]
So here's what I don't quite get.

The windows code at startup does this

```cpp
  const auto installPath = sst::plugininfra::paths::sharedLibraryBinaryPath().parent_path();

...

    auto cp{installPath};

    while (userDataPath.empty() && cp.has_parent_path() && cp != cp.parent_path())
    {
        auto portable = cp / L"SurgeXTUserData";

        if (fs::is_directory(portable))
        {
            userDataPath = std::move(portable);
        }

        cp = cp.parent_path();
    }

    if (userDataPath.empty())
    {
        userDataPath = sst::plugininfra::paths::bestDocumentsFolderPathFor(sxt);
    }
```

This means if you have a `SurgeXTUserData` junction in the parent path of your VST it will work. You can point that wherever you want.

@mkruselj isn't that just the solution here?

The line which throws the exception on th misconfigured system is the `bestDocumentsFolderPathFor(sxt)`

we could also try-catch that but then we don't have enough surge to pop a ui yet.

But if creating the SurgeXTUserData is the solution we could also special case the fatal error for this documents folder path.

(Of course, windows machines should answer the documents api with a path! this is still a misconfigured underlying system).

**Comment by mkruselj** [comment]
Symlink could def be created but it would carry the access rights of the target. So if target is readable but not writable we're still in a bad situation.

**Comment by baconpaul** [comment]
RIght

So I'm breaking my mac install now (`userDataPath=fs::path{"/not/writable"}` and have surge starting but pretty broken. If I can get it to pop a non-fatal error in this case I'll push that in the windows branch. Stay tuned.

**Comment by baconpaul** [comment]
OK so I was able to get surge partly running with a different non-fatal error message when the path is broken on my mac. Some things just wont work - basically anything that requires a write to the location. But you can at least play presets and stuff.

If 7552 passes CI I'll merge that and then you can try a nightly.

**Comment by baconpaul** [comment]
> Symlink could def be created but it would carry the access rights of the target. So if target is readable but not writable we're still in a bad situation.

I meant a solution for the user here is that they create the symlink to a writable place, though. That should work right?

**Comment by mkruselj** [comment]
Yup that is correct.

**Comment by baconpaul** [comment]
Ok I merged a change which will at least not make the error fatal. @parduz a new nightky will be at https://surge-synthesizer.github.io/nightly_XT/ soon enough 

the fix toh can use even after that is to find the junction named surgextuserdata which should be next to your vst3 and point it some place writable. Or fix your config so documents folder is resolvable

**Comment by parduz** [comment]
> Ok I merged a change which will at least not make the error fatal. @parduz a new nightky will be at https://surge-synthesizer.github.io/nightly_XT/ soon enough
> 
> the fix toh can use even after that is to find the junction named surgextuserdata which should be next to your vst3 and point it some place writable. Or fix your config so documents folder is resolvable

@baconpaul ,i have not understood:
It wasn't me with the problem to the non writable path.... and i won't make my document folder non accessible
were you asking me to do some tests?

Sorry, my english is bad

**Comment by baconpaul** [comment]
If you get that nightky surge will start and you can play presets but lots of things like preferences won’t work

if you want that to work your options are 

1. find the hard link called “surgextuserdata” in your install folder and point it at a writable directory or
2. Reset the location of your documents folder to something writable in the registry somehow. 

Users have had success with number 2

my change was to make the fatal error a bit less fatal but still not resolved

we are having a convo about potentially using appdata as a third choice if documents is missing in the developer group which would be another solution for MIs configured systems

Hope that is clear! Please ask if not. Thanks

**Comment by baconpaul** [comment]
@parduz have you had a chance to test this change?

the error message you got in the original post is “directory not found” which the function should in theory never return. We will recover more gracefully now though at least

**Comment by mkruselj** [comment]
@NuloSois is OP who should report results of latest nightly, actually.

**Comment by baconpaul** [comment]
Oh I'm sorry!
